lib/linked_list.rb

```diff
@@ -21,39 +21,85 @@ def initialize
     # Time Complexity:  
     # Space Complexity
     def add_first(value)
-      raise NotImplementedError
+      new_node = Node.new(value)
+
+      if @head 
+        new_node.next = @head
+      end
+
+      @head = new_node
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     # Time Complexity:  
     # Space Complexity
     def search(value)
-      raise NotImplementedError
+      current = @head
+      while current
+        if current.data == value
+          return true
+        end
+        current = current.next
+      end
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     # Time Complexity:  
     # Space Complexity
     def find_max
-      raise NotImplementedError
+      if !@head
+        return nil
+      end
+      
+      current = @head
+      max = current.data
+      current = current.next
+      while current
+        if current.data > max
+        max = current.data
+      end
+      current = current.next
+    end
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
     # Time Complexity:  
     # Space Complexity
     def find_min
-      raise NotImplementedError
+      if !@head
+        return nil
+      end
+      
+      current = @head
+      min = current.data
+      current = current.next
+      while current
+        if current.data < min
+        min = current.data
+      end
+      current = current.next
+    end
+      return min
     end
 
 
     # method that returns the length of the singly linked list
     # Time Complexity:  
     # Space Complexity
     def length
-      raise NotImplementedError
+      i = 0
+      current = @head
+      
+      while current
+        i++

```

Comment:
++ doesn't really work as an operator in Ruby.  It treats this as i plus some positive number.

```suggestion
        i += 1
```